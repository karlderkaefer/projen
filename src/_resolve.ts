import { ResolveOptions, IResolvable } from "./file";

function isResolvable(obj: any): obj is IResolvable {
  return (obj as IResolvable).toJSON !== undefined;
}

export function resolve(value: any, options: ResolveOptions = {}): any {
  const args = options.args ?? [];
  const omitEmpty = options.omitEmpty ?? false;

  if (value == null) {
    return value;
  }

  if (isResolvable(value)) {
    const resolved = value.toJSON();
    return resolve(resolved, options);
  }

  // if value is a function, call it and resolve the result.
  if (typeof value === "function") {
    const resolved = value.apply(undefined, args);
    return resolve(resolved, options);
  }

  if (typeof value !== "object") {
    return value;
  }

  if (Array.isArray(value)) {
    if (omitEmpty && value.length === 0) {
      return undefined;
    }
    return value.map((x) => resolve(x, options)).filter((x) => x != null); // filter undefined/null/omitted
  }

  // only allow data types (i.e. objects without constructors)
  if (value.constructor && value.constructor?.name !== "Object") {
    throw new Error(
      `only data types can be resolved. trying to resolve object of type ${value.constructor?.name}`
    );
  }

  const result: any = {};

  for (const [k, v] of Object.entries(value)) {
    const resolved = resolve(v, options);

    // skip undefined values
    if (resolved === undefined) {
      continue;
    }

    result[k] = resolved;
  }

  if (omitEmpty && Object.keys(result).length === 0) {
    return undefined;
  }

  return result;
}
