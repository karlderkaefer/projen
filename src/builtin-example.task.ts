console.log("hello, I am a builtin task");
console.log("second line");
