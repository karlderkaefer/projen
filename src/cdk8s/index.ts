export * from "./auto-discover";
export * from "./cdk8s-app-ts";
export * from "./cdk8s-construct";
export * from "./integration-test";
