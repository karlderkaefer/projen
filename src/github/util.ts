export function secretToString(secretName: string): string {
  return `\${{ secrets.${secretName} }}`;
}
