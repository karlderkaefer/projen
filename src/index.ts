export * from './awscdk-app-ts';
export * from './awscdk-construct';
export * from './cdk8s-app-ts';
export * from './cdk8s-construct';
export * from './cdktf-construct';
export * from './component';
export * from './construct-lib';
export * from './dev-env';
export * from './docker-compose';
export * from './eslint';
export * from './file';
export * from './gitpod';
export * from './ignore-file';
export * from './ini';
export * from './jest';
export * from './jsii-project';
export * from './json';
export * from './logger';
export * from './license';
export * from './makefile';
export * from './node-package';
export * from './node-project';
export * from './object-file';
export * from './option-hints';
export * from './project';
export * from './readme';
export * from './sample-file';
export * from './semver';
export * from './textfile';
export * from './toml';
export * from './typescript';
export * from './typescript-config';
export * from './version';
export * from './yaml';
export * from './xmlfile';
export * from './source-code';
export * from './upgrade-dependencies';
export * from './circle-ci';

// export submobules
export * as tasks from './tasks';
export * as web from './web';
export * as github from './github';
export * as vscode from './vscode';
export * as deps from './deps';
export * as json from './json/index'; // <-- we also have json.ts
export * as java from './java';
export * as python from './python';
export * as typescript from './typescript/index'; // <-- we also have typescript.ts
export * as javascript from './javascript';
export * as release from './release';
export * as git from './git';
