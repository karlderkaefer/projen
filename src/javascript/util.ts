import { basename, dirname, extname, join, sep } from "path";

export function renderBundleName(entrypoint: string) {
  const parts = join(entrypoint).split(sep);
  if (parts[0] === "src") {
    parts.shift(); // just remove 'src' if its the first element for ergonomics
  }

  const p = parts.join(sep);
  const dir = dirname(p);
  const base = basename(p, extname(p));
  return join(dir, base);
}
