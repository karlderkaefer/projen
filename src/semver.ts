import * as semver from "semver";

/**
 * @deprecated This class will be removed in upcoming releases. if you wish to
 * specify semver requirements in `deps`, `devDeps`, etc, specify them like so
 * `express@^2.1`.
 */
export class Semver {
  public static of(spec: string) {
    return new Semver(spec);
  }

  /**
   * Latest version.
   */
  public static latest() {
    return new Semver("*");
  }

  /**
   * Accept only an exact version
   */
  public static pinned(version: string) {
    return new Semver(version);
  }

  /**
   * Accept any minor version.
   *
   * >= version
   * < next major version
   */
  public static caret(version: string) {
    return new Semver(`^${version}`);
  }

  /**
   * Accept patches.
   *
   * >= version
   * < next minor version
   */
  public static tilde(version: string) {
    return new Semver(`~${version}`);
  }

  public readonly mode?: string;

  private constructor(public readonly spec: string) {}

  public get version() {
    return semver.minVersion(this.spec)?.version;
  }
}
