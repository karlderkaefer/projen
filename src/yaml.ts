import * as YAML from "yaml";
import { IResolver } from "./file";
import { ObjectFile, ObjectFileOptions } from "./object-file";
import { Project } from "./project";

/**
 * Options for `JsonFile`.
 */
export interface YamlFileOptions extends ObjectFileOptions {
  /**
   * Maximum line width (set to 0 to disable folding).
   *
   * @default - 0
   */
  readonly lineWidth?: number;
}

/**
 * Represents a YAML file.
 */
export class YamlFile extends ObjectFile {
  /**
   * Maximum line width (set to 0 to disable folding).
   */
  public lineWidth: number;

  constructor(project: Project, filePath: string, options: YamlFileOptions) {
    super(project, filePath, options);
    this.lineWidth = options.lineWidth ?? 0;
  }

  protected synthesizeContent(resolver: IResolver): string | undefined {
    const json = super.synthesizeContent(resolver);
    if (!json) {
      return undefined;
    }

    return [
      ...(this.marker ? [`# ${this.marker}`] : []),
      "",
      YAML.stringify(JSON.parse(json), {
        indent: 2,
        lineWidth: this.lineWidth,
      }),
    ].join("\n");
  }
}
