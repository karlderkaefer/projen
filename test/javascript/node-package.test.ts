import { readFileSync, writeFileSync, mkdirSync } from "fs";
import { join } from "path";
import { readJsonSync } from "fs-extra";
import semver from "semver";
import * as YAML from "yaml";
import { Project } from "../../src";
import {
  NodePackage,
  NodePackageManager,
} from "../../src/javascript/node-package";
import { minVersion } from "../../src/javascript/util";
import * as util from "../../src/util";
import { mkdtemp, synthSnapshot, TestProject } from "../util";

/**
 * Mocks a yarn install, writing to yarn.lock
 * and creating package.json files in node_modules for all dependencies
 * Will "install" the max version that complies with all dependency ranges
 * NOT A PERFECT MODEL OF YARN. JUST CLOSE ENOUGH.
 * @param outdir Test project's outdir, where package.json and node_modules live
 * @param latestPackages Package name and version to "install" for "*" deps
 */
function mockYarnInstall(
  outdir: string,
  latestPackages: Record<string, string>
) {
  const pkgJson = readJsonSync(join(outdir, "package.json"));
  const yarnLock: Record<string, string> = {};
  const depRanges: Record<string, string[]> = {};
  const depVersions: Record<string, string[]> = {};
  const depTypes = [
    "dependencies",
    "devDependencies",
    "peerDependencies",
    "bundledDependencies",
    "optionalDependencies",
  ];
  for (const depType of depTypes) {
    // Look up the ranges for each dependency
    for (const [depName, depVer] of Object.entries(pkgJson[depType] ?? {})) {
      // init arrays if undefined
      depRanges[depName] = depRanges[depName] ?? [];
      depVersions[depName] = depVersions[depName] ?? [];
      if (depVer === "*") {
        if (!latestPackages[depName]) {
          throw new Error(
            `No latest package defined for dependency ${depName}`
          );
        }
        depRanges[depName].push("*");
        depVersions[depName].push(latestPackages[depName]);
      } else {
        const ver = minVersion(`${depVer}`);
        if (!ver) {
          throw new Error(
            `unable to determine min version for ${depName}@${depVer}`
          );
        }
        // If we're given a latest package, no dependency can require higher
        if (
          latestPackages[depName] &&
          semver.validRange(`${ver}`) &&
          semver.gt(`${ver}`, latestPackages[depName])
        ) {
          throw new Error(
            `${depType} requirement ${depName}@${depVer} exceeds provided "latest" version ${latestPackages[depName]}`
          );
        }
        depRanges[depName].push(`${depVer}`);
        // Also take the min version as a valid install version
        depVersions[depName].push(ver);
      }
    }
  }
  // Resolve version to install that satisfies all ranges for dep
  for (const dep of Object.keys(depRanges)) {
    let installVersion: string | null;
    if (!semver.validRange(depVersions[dep][0])) {
      // if a dependency is "file:..." or something else, just install that
      installVersion = depVersions[dep][0];
    } else {
      installVersion = semver.maxSatisfying(
        depVersions[dep],
        depRanges[dep].join(" || ")
      );
    }
    if (!installVersion) {
      throw new Error(`No version given satisfies all constraints on ${dep}`);
    }
    mkdirSync(join(outdir, `node_modules/${dep}`), { recursive: true });
    writeFileSync(
      join(outdir, `node_modules/${dep}/package.json`),
      JSON.stringify({
        name: `${dep}`,
        version: `${installVersion}`,
      })
    );
    // Not accurate to yaml.lock v1 format, but close enough.
    yarnLock[
      `${depRanges[dep].map((range) => `${dep}@${range}`).join(", ")}`
    ] = `version "${installVersion}"`;
  }
  // Use double quoted keys just to make output more predictable
  writeFileSync(
    join(outdir, "yarn.lock"),
    YAML.stringify(yarnLock, { defaultKeyType: "QUOTE_DOUBLE" })
  );
}

afterEach(() => {
  jest.resetAllMocks();
  jest.restoreAllMocks();
});

test("all bugs field present", () => {
  const project = new TestProject();

  new NodePackage(project, {
    bugsEmail: "bugs@foobar.local",
    bugsUrl: "bugs.foobar.local",
  });

  expect(synthSnapshot(project)["package.json"].bugs).toMatchSnapshot();
});

test("no bugs field present", () => {
  const project = new TestProject();

  new NodePackage(project, {});

  const snps = synthSnapshot(project);

  expect(snps["package.json"].bugs).toMatchSnapshot();

  expect(snps["package.json"].bugs).toStrictEqual(undefined);
});

test("single bugs field present", () => {
  const project = new TestProject();

  const _email = "bugs@foobar.local";

  new NodePackage(project, {
    bugsEmail: _email,
  });

  const snps = synthSnapshot(project);

  expect(snps["package.json"].bugs).toMatchSnapshot();

  expect(snps["package.json"].bugs.url).toStrictEqual(undefined);
  expect(snps["package.json"].bugs.email).toStrictEqual(_email);
});

test('lockfile updated (install twice) after "*"s are resolved', () => {
  const execMock = jest
    .spyOn(util, "exec")
    .mockImplementation((command, options) => {
      expect(command.startsWith("yarn install")).toBeTruthy();
      mockYarnInstall(options.cwd, { ms: "2.1.3" });
    });

  const project = new Project({ name: "test" });
  const pkg = new NodePackage(project);

  pkg.addDeps("ms");

  project.synth();

  const yarnLockPath = join(project.outdir, "yarn.lock");
  const yarnLock: string | undefined = readFileSync(yarnLockPath, "utf8");

  expect(yarnLock).toStrictEqual('"ms@^2.1.3": version "2.1.3"\n');
  expect(execMock).toBeCalledTimes(2);
});

test("install only once if all versions are resolved", () => {
  const execMock = jest.spyOn(util, "exec").mockReturnValueOnce();
  const project = new Project({ name: "test" });
  const pkg = new NodePackage(project);

  pkg.addDeps("ms@^2");

  project.synth();

  expect(execMock).toBeCalledTimes(1);
});

test("no install if package.json did not change at all", () => {
  const execMock = jest.spyOn(util, "exec").mockReturnValueOnce();
  const outdir = mkdtemp({ cleanup: false });

  const orig = {
    name: "test",
    scripts: {
      build: "npx projen build",
      compile: "npx projen compile",
      default: "npx projen default",
      eject: "npx projen eject",
      package: "npx projen package",
      "post-compile": "npx projen post-compile",
      "pre-compile": "npx projen pre-compile",
      test: "npx projen test",
    },
    dependencies: {
      ms: "^2",
    },
    main: "lib/index.js",
    license: "Apache-2.0",
    version: "0.0.0",
    "//": '~~ Generated by projen. To modify, edit .projenrc.js and run "npx projen".',
  };

  writeFileSync(
    join(outdir, "package.json"),
    JSON.stringify(orig, undefined, 2)
  );
  mkdirSync(join(outdir, "node_modules")); // <-- also causes an "install"

  const project = new Project({ name: "test", outdir });
  project.addExcludeFromCleanup("package.json");
  const pkg = new NodePackage(project);

  pkg.addDeps("ms@^2");

  project.synth();
  expect(execMock).not.toBeCalled();
});

test('"*" peer dependencies are pinned in devDependencies', () => {
  // Post-synth dependency version resolution uses installed package from node_modules folder
  // Mock install command to add this folder with a fixed dependency version,
  // mimicking yarn installing the latest package for "*"
  jest.spyOn(util, "exec").mockImplementation((command, options) => {
    expect(command.startsWith("yarn install")).toBeTruthy();
    mockYarnInstall(options.cwd, { ms: "1.2.3" });
  });

  const project = new Project({ name: "test" });
  const pkg = new NodePackage(project, {
    peerDependencyOptions: {
      pinnedDevDependency: true,
    },
  });

  pkg.addPeerDeps("ms");

  project.synth();

  const pkgFile = readJsonSync(join(project.outdir, "package.json"));

  expect(pkgFile.peerDependencies).toStrictEqual({ ms: "^1.2.3" });
  expect(pkgFile.devDependencies).toStrictEqual({ ms: "1.2.3" });
});

test("manually set devDependencies are not changed when a peerDependency is added", () => {
  // Post-synth dependency version resolution uses installed package from node_modules folder
  // Mock install command to add this folder with a fixed dependency version,
  // mimicking yarn installing the latest package for "*"
  jest.spyOn(util, "exec").mockImplementation((command, options) => {
    expect(command.startsWith("yarn install")).toBeTruthy();
    mockYarnInstall(options.cwd, { ms: "1.3.4" });
  });

  const project = new Project({ name: "test" });
  const pkg = new NodePackage(project, {
    peerDependencyOptions: {
      pinnedDevDependency: true,
    },
  });

  const orig = {
    name: "test",
    devDependencies: {
      ms: "^1.3.0",
    },
    main: "lib/index.js",
    license: "Apache-2.0",
    version: "0.0.0",
    "//": '~~ Generated by projen. To modify, edit .projenrc.js and run "npx projen".',
  };

  writeFileSync(
    join(project.outdir, "package.json"),
    JSON.stringify(orig, undefined, 2)
  );

  pkg.addPeerDeps("ms");

  project.synth();

  const pkgFile = readJsonSync(join(project.outdir, "package.json"));

  expect(pkgFile.peerDependencies).toStrictEqual({ ms: "^1.3.4" });
  expect(pkgFile.devDependencies).toStrictEqual({ ms: "^1.3.0" });
});

test("devDependencies are not pinned by peerDependencies if a regular (runtime) dependency also exists", () => {
  // Post-synth dependency version resolution uses installed package from node_modules folder
  // Mock install command to add this folder with a fixed dependency version,
  // mimicking yarn installing the latest package for "*"
  jest.spyOn(util, "exec").mockImplementation((command, options) => {
    expect(command.startsWith("yarn install")).toBeTruthy();
    mockYarnInstall(options.cwd, { ms: "1.3.8" });
  });

  const project = new Project({ name: "test" });
  const pkg = new NodePackage(project, {
    peerDependencyOptions: {
      pinnedDevDependency: true,
    },
  });

  pkg.addDeps("ms");
  pkg.addPeerDeps("ms");

  project.synth();

  const pkgFile = readJsonSync(join(project.outdir, "package.json"));

  expect(pkgFile.peerDependencies).toStrictEqual({ ms: "^1.3.8" });
  expect(pkgFile.dependencies).toStrictEqual({ ms: "^1.3.8" });
  expect(pkgFile.devDependencies).toBeUndefined();
});

test("devDependencies are not pinned by peerDependencies if pinnedDevDependency is false", () => {
  // Post-synth dependency version resolution uses installed package from node_modules folder
  // Mock install command to add this folder with a fixed dependency version,
  // mimicking yarn installing the latest package for "*"
  jest.spyOn(util, "exec").mockImplementation((command, options) => {
    expect(command.startsWith("yarn install")).toBeTruthy();
    mockYarnInstall(options.cwd, { ms: "1.4.0" });
  });

  const project = new Project({ name: "test" });
  const pkg = new NodePackage(project, {
    peerDependencyOptions: {
      pinnedDevDependency: false,
    },
  });

  pkg.addPeerDeps("ms");

  project.synth();

  const pkgFile = readJsonSync(join(project.outdir, "package.json"));

  expect(pkgFile.peerDependencies).toStrictEqual({ ms: "^1.4.0" });
  expect(pkgFile.devDependencies).toBeUndefined();
});

test("file path dependencies are respected", () => {
  // Post-synth dependency version resolution uses installed package from node_modules folder
  // Mock install command to add this folder with a fixed dependency version,
  // mimicking yarn installing the latest package for "*"
  jest.spyOn(util, "exec").mockImplementation((command, options) => {
    expect(command.startsWith("yarn install")).toBeTruthy();
    mockYarnInstall(options.cwd, { ms: "file:../ms" });
  });

  const project = new Project({ name: "test" });
  const pkg = new NodePackage(project, {
    peerDependencyOptions: {
      pinnedDevDependency: false,
    },
  });

  pkg.addPeerDeps("ms@file:../ms");

  project.synth();

  const pkgFile = readJsonSync(join(project.outdir, "package.json"));

  expect(pkgFile.peerDependencies).toStrictEqual({ ms: "file:../ms" });
  expect(pkgFile.devDependencies).toBeUndefined();
});

test("local dependencies can be specified using 'file:' prefix", () => {
  jest.spyOn(util, "exec");
  const localDepPath = mkdtemp({ cleanup: false });
  const localPackage = {
    name: "local-dep",
    version: "0.0.0",
  };

  writeFileSync(
    join(localDepPath, "package.json"),
    JSON.stringify(localPackage, undefined, 2)
  );

  const project = new TestProject();
  const pkg = new NodePackage(project);
  pkg.addPeerDeps(`file:${localDepPath}`);

  project.synth();

  const pkgFile = readJsonSync(join(project.outdir, "package.json"));

  expect(pkgFile.peerDependencies).toStrictEqual({ "local-dep": localDepPath });
});

test("yarn resolutions", () => {
  const project = new TestProject();

  const pkg = new NodePackage(project, {
    packageManager: NodePackageManager.YARN,
  });

  pkg.addPackageResolutions("some-dep@1.0.0", "other-dep");

  const snps = synthSnapshot(project);

  expect(snps["package.json"].resolutions).toBeDefined();
  expect(snps["package.json"]).toMatchSnapshot();
});

test("npm overrides", () => {
  const project = new TestProject();

  const pkg = new NodePackage(project, {
    packageManager: NodePackageManager.NPM,
  });

  pkg.addPackageResolutions("some-dep@1.0.0", "other-dep");

  const snps = synthSnapshot(project);

  expect(snps["package.json"].overrides).toBeDefined();
  expect(snps["package.json"]).toMatchSnapshot();
});

test("pnpm overrides", () => {
  const project = new TestProject();

  const pkg = new NodePackage(project, {
    packageManager: NodePackageManager.PNPM,
  });

  pkg.addPackageResolutions("some-dep@1.0.0", "other-dep");

  const snps = synthSnapshot(project);

  expect(snps["package.json"].pnpm.overrides).toBeDefined();
  expect(snps["package.json"]).toMatchSnapshot();
});
