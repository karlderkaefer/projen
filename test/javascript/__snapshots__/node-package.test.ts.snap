// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`all bugs field present 1`] = `
Object {
  "email": "bugs@foobar.local",
  "url": "bugs.foobar.local",
}
`;

exports[`no bugs field present 1`] = `undefined`;

exports[`single bugs field present 1`] = `
Object {
  "email": "bugs@foobar.local",
}
`;
